### PR TITLE
Various fixes and additional code

### PR DESCRIPTION
**Pull request by budiariyanto**
- JWT middleware
- Log middleware
- Finalize log
- Custom HTTP handler
- Custom HTTP writer and error writer
- JWT lib

### CODE REVIEW

File: middleware/jwt.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	return &JWTMiddleware{
+		signKey: signKey,
+		jwt:     jwt.NewJWT(signKey),
+		CustomWriter: phttp.CustomWriter{
+			C: hctx,
+		},
+	}
+}
+
+// func NewHandlerContext()
```
Comment (cikupin) [inline middleware/jwt.go]: dihapus aja mas

Comment (budiariyanto) [reply middleware/jwt.go]: wah iya

File: middleware/jwt.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+		return
+	}
+
+	parentCtx := r.Context()
+	ctx := context.WithValue(parentCtx, "token", claims)
+
+	next(w, r.WithContext(ctx))
+}
+
+// func (m *Middleware) IsTokenExpired(err error) bool {
```
Comment (cikupin) [inline middleware/jwt.go]: dihapus aja mas

File: middleware/log.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+func NewHttpRequestLoggerMiddleware(logger *log.Logger) *HttpRequestLoggerMiddleware {
+	return &HttpRequestLoggerMiddleware{
+		logger: logger,
+	}
+}
+
+func (l *HttpRequestLoggerMiddleware) ServeHTTP(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	l.logger.SetRequest(r)
+	next(w, r)
```
Comment (cikupin) [inline middleware/log.go]: logger.SetResponse() nya gak ada mas ?

File: token/jwt/jwt.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	tkn, err := j.Parse(token)
+	if err != nil {
+		return
+	}
+
+	valid = tkn.Valid
+	return
+}
+
+// func (m *Middleware) IsTokenExpired(err error) bool {
```
Comment (cikupin) [inline token/jwt/jwt.go]: ini juga dihapus aja mas

File: middleware/log.go (hunk 3 above)
Comment (budiariyanto) [reply middleware/log.go]: set responsenya dipanggil di handler masing2

File: token/jwt/jwt_test.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	assert.Nil(t, err, "Error must be nil 1")
+
+	parsedToken, err := jwt.Parse(token)
+	assert.Nil(t, err, err)
+	t.Log(parsedToken.Header)
+	t.Log(parsedToken.Claims)
+	t.Log(parsedToken.Signature)
+}
+
+// func TestValid(t *testing.T) {
```
Comment (cikupin) [inline token/jwt/jwt_test.go]: dihapus juga deh mas

**Pull request merged**